addons/cookoff/XEH_postInit.sqf

```diff
@@ -58,3 +58,50 @@
         [FUNC(detonateAmmunition), [_vehicle, _mags, _total], _delay] call CBA_fnc_waitAndExecute;
     };
 }, nil, ["Man","StaticWeapon"]] call CBA_fnc_addClassEventHandler;
+
+if (hasInterface) then {
+    GVAR(soundObjPool) = [];
+    [QFUNC(poolGetObject), {
+        private _pooledObject = GVAR(soundObjPool) param [0, [-999, objNull]];
+        _pooledObject params ["_time", "_obj"];
+        if (isNull _obj || _time >= time) then {
+            _obj = "#particlesource" createVehicleLocal [0, 0, 0];
+        } else {
+            GVAR(soundObjPool) deleteAt 0;
+        };
+        GVAR(soundObjPool) pushBack [time + 8, _obj];
+        _obj; // return
+    }] call CBA_fnc_compileFinal;
+
+    [QGVAR(playCookoffSound), {
+        params ["_obj", "_sound", "_maxDist"];
+        if (isNull _obj) exitWith {};
+        private _distance = _obj distance (positionCameraToWorld [0,0,0]);
+        if (_distance > _maxDist) exitWith {};
+
+        // Delay sound after Rule of SOS.
+        [{
+            params ["_obj", "_sound", "_maxDist"];
+            private _distance = _obj distance (positionCameraToWorld [0, 0, 0]); // update Value so that distance is correct to the current camera Position
+            private _sDistance = _distance + ((random 5) - 2.5);
+            private _disStep = [[["far", 1.5], ["mid", 1.3]] select (_sDistance < 952), ["close", 1]] select (_sDistance < 235);
+            _sound = format [QGVAR(%1_%2_%3), _sound, _disStep select 0, ceil (random 3)];
+            hintSilent _sound;
+            // fix for Vanilla Say3d Falloff
+            private _value = 1 - (_distance / _maxDist);
+            private _fakeValue = linearConversion ([[0, 0.7, _value, 1, 0.8], [0.7, 1, _value, 0.8, 0]] select (_value >= 0.7));
+            private _soundSource = call FUNC(poolGetObject);
+            _soundSource setPos (getPos _obj);
+            _soundSource say3D [_sound, (_distance / _fakeValue) * (_disStep select 1), 0.9 + (random 0.2)];
+        }, [_obj, _sound, _maxDist], _distance / SPEED_OF_SOUND] call CBA_fnc_waitAndExecute;
+    }] call CBA_fnc_addEventHandler;
+
+    [{
+        if (GVAR(soundObjPool) isEqualTo []) exitWith {};
+        private _element = GVAR(soundObjPool) param [0, [time, objNull]];
+        if (((_element select 0) + 120) < time) then {
+            GVAR(soundObjPool) deleteAt 0;
+            deleteVehicle (_element select 1);
+        };
+    }, 0.5] call CBA_fnc_addPerFrameHandler;
```

Comment:
I don't know much about the calculations behind sound waves, but my suggestion is as follows:

```suggestion
    // Plays a sound locally, so that different sounds can be used for various distances
    [QGVAR(playCookoffSound), {
        params ["_object", "_sound", "_maxDistance"];

        if (isNull _object) exitWith {};

        private _distance = _object distance (positionCameraToWorld [0, 0, 0]);

        if (_distance > _maxDistance) exitWith {};

        TRACE_3("",_object,_sound,_maxDistance);

        // 3 classes of distances: close, mid and far, each having different sound files
        _distance = ([["far", "mid"] select (_distance < 952), "close"] select (_distance < 235));
        _sound = format ["sounds\%1\%2_%3.wss", _sound, _distance, floor (random 3) + 1];

        TRACE_1("",_sound);

        _sound = FORMAT_1(QPATHTOF(%1),_sound);

        // Make sure file exists, so RPT isn't spammed with non-existent entry errors
        if (!fileExists _sound) exitWith {};

        // Obeys speed of sound and, according to wiki, takes doppler effects into account
        playSound3D [_sound, objNull, insideBuilding _object >= 0.5, getPosASL _object, 1, 0.9 + (random 0.2), _maxDistance, 0, true];
    }] call CBA_fnc_addEventHandler;
```

It's a very basic implementation which I detailed in https://github.com/acemod/ACE3/issues/9547#issuecomment-1799740572.

I dropped the fall-off calculations made in the PR, as `playSound3D` already accounts for that. I also removed the volume calculations, as I did not know how to re-implement them into my suggestion.

I don't really have the knowledge and setup to judge if the quality is acceptable, so if anyone knowledgeable could find calculations for the volume (and fall-off if necessary) that would be much appreciated.

Reply:
This does not cover the shotmissile and shotsubmunitions. those where done via a https://github.com/acemod/ACE3/pull/5327/files#diff-73a0f49fbad6f9b061d350aa515eef040577f4c7b959bda32d5c654c643651d3R24-R42

Reply:
I have updated my suggestion to include `shotmissile` and `shotsubmunitions`.